mediatek: add support for Ruijie RG-X60 Pro

NekokeCore opened the pull request:
This pull request ports Ruijie RG-X60 Pro router support to the main branch.

# Parameters
```
- SoC       : MediaTek MT7986A Quad-core ARM® Cortex-A53 2.0GHz
- RAM       : DDR3 512MiB (W634GU6QB)
- Flash     : SPI-NAND 128 MiB (W25N01GVZEIG)
- WLAN      : MediaTek MT7986A integration dual-band WiFi 6
  - 2.4 GHz : b/g/n/ax, MIMO 4x4
  - 5 GHz   : a/n/ac/ax, MIMO 4x4
- Ethernet  : 10/100/1000 Mbps x4 (MediaTek MT7531AE) 
                    2500Mbps x 1 （Realtek RTL8221B-VB-CG）
- UART      : through-hole on PCB
  - [J500] GND, TX, RX, 3.3V (115200n1)
- Buttons   : Mesh, Reset
- LEDs      : 1x Power (Blue)
                  1x Turbo (Purple)
- Power     : 12 VDC, 3 A
```

# How to Installation
1. Remove the case and connect the TTL cable to the corresponding position.
2. Power on the device and quickly press "down" on the keyboard, then U-Boot will stay in the menu.
3. Select "1. Upgrade Firmware", select "0. TFTP Client(Default)".
4. Input the IP address, input the Openwrt image file name to be flashed, start the TFTP server, and press "Enter".
5. Wait for the flashing to complete.

# How return to stock
1. Remove the case and connect the TTL cable to the corresponding position.
2. Power on the device and quickly press "down" on the keyboard, then U-Boot will stay in the menu.
3. Select "1. Upgrade Firmware", select "0. TFTP Client(Default)".
4. Input the IP address, input the Stock “E-WEBOS” image file name to be flashed, start the TFTP server, and press "Enter".
5. Wait for the flashing to complete.

# About recovery
Connect uart, use u-boot menu to flash stock firmware image or boot OpenWrt initramfs image.

# About MAC Address
```
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 10:82:3D:XX:XX:9E | label     |
| LAN     | 10:82:3D:XX:XX:9F | label+1   |
| WLAN 2g | 10:82:3D:XX:XX:A0 | label+2   |
| WLAN 5g | 10:82:3D:XX:XX:A1 | label+3   |
+---------+-------------------+-----------+
```

mrkiko commented:
Hi!

first of all, thank you for your work.
There are some formal issues still: you need to use your real name and add a valid e-mail address in the Signed-off-by tag.
Furthermore, did you test sysupgrade on this device? Does it work?
Then, you may wish to include some installation / recovery instructions.

Thanks,
Enrico

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,432 @@
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		reset {
```
Please, follow the pattern https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/input/gpio-keys.yaml#L26. 
e.g. `button-0`

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,432 @@
+	leds-gpio {
+		compatible = "gpio-leds";
+
+		sysled: sysled {
```
Please follow the `"(^led-[0-9a-f]$|led)"` pattern. e.g. `led-0`, `led-1`

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,432 @@
+		compatible = "gpio-leds";
+
+		sysled: sysled {
+			label = "white:system";
```
New style would be better instead of the label.

```
			color = <LED_COLOR_ID_WHITE>;
			function = LED_FUNCTION_STATUS;
```
https://github.com/torvalds/linux/blob/master/include/dt-bindings/leds/common.h

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,432 @@
+			factory: partition@180000 {
+				label = "Factory";
+				reg = <0x180000 0x200000>;
+				phandle = <0x17>;
```
Why do you need this `phandle`?

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,432 @@
+				read-only;
+			};
+
+			factory: partition@180000 {
```
Would be nice to define the eeprom here using nvmem-layout.
https://github.com/openwrt/openwrt/pull/14054/files#diff-5b63e767546137b52411c6122b35e1882837aea6a0fb817d99df703a8bf982f5R173

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -52,6 +52,10 @@ zyxel,ex5601-t0-ubootmod)
+	ucidef_set_led_default "power" "power" "white:system" "0"
+	ucidef_set_led_default "trubo" "trubo" "purple:trubo" "1"
+	;;
 esac
```
Please sort this block alphabetically,

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -82,6 +82,10 @@ mediatek_setup_interfaces()
 	zyxel,ex5601-t0-ubootmod)
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" eth1
 		;;
+	ruijie,ruijie-rg-x60-pro)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" eth1
+		;;
+
```
Sorting. And this block can be merged with the previous one.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -164,6 +168,11 @@ mediatek_setup_macs()
 		wan_mac=$(mtd_get_mac_ascii Bdata ethaddr_wan)
 		label_mac=$wan_mac
 		;;
+	ruijie,ruijie-rg-x60-pro)
+		wan_mac=$(mtd_get_mac_ascii Bdata ethaddr_wan)
+		label_mac=$wan_mac
+		;;
+
```
The same as above.

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -942,3 +942,17 @@ define Device/zyxel_nwa50ax-pro
+  DEVICE_MODEL := Ruijie RG X60 Pro
+  DEVICE_DTS := mt7986a-ruijie-rg-x60-pro
+  DEVICE_DTS_DIR := ../dts
+  UBINIZE_OPTS := -E 5
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL_IN_UBI := 1
```
Do you need this lines? I assume they can be dropped.

csharper2005 replied:
Still actual.

csharper2005 reviewed (commented):
And... Please MAC address table to the commit message.

NekokeCore commented:
> And... Please MAC address table to the commit message.

Thanks for pointing these out, I'll correct them later!

NekokeCore commented:
> Hi!
> 
> first of all, thank you for your work. There are some formal issues still: you need to use your real name and add a valid e-mail address in the Signed-off-by tag. Furthermore, did you test sysupgrade on this device? Does it work? Then, you may wish to include some installation / recovery instructions.
> 
> Thanks, Enrico

It‘s worked,all radio ,1 x 2.5GbE,4x 1GbE are fine. I'll fill in the missing information one by one.

<img width="895" alt="f64345bf1e3423cda6947a51c1140cd6" src="https://github.com/openwrt/openwrt/assets/42233090/e1b0c261-1a67-4f6b-96dc-2445e089944d">

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -16,7 +16,8 @@ mediatek_setup_interfaces()
 	mediatek,mt7981-rfb|\
-	zbtlink,zbt-z8102ax)
+	zbtlink,zbt-z8102ax|\
+	ruijie,ruijie-rg-x60-pro)
```
Sorting.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -82,6 +83,7 @@ mediatek_setup_interfaces()
 	zyxel,ex5601-t0-ubootmod)
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" eth1
 		;;
+
```
Blank line can be dropped.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -160,10 +162,12 @@ mediatek_setup_macs()
 	xiaomi,redmi-router-ax6000-stock|\
-	xiaomi,redmi-router-ax6000-ubootmod)
+	xiaomi,redmi-router-ax6000-ubootmod|\
+	ruijie,ruijie-rg-x60-pro)
```
Please sort them alphabetically.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -160,10 +162,12 @@ mediatek_setup_macs()
 		wan_mac=$(mtd_get_mac_ascii Bdata ethaddr_wan)
 		label_mac=$wan_mac
 		;;
+
```
Can be dropped.

csharper2005 reviewed (commented):
You need to squash your commits into the one:
https://github.com/openwrt/openwrt/pull/13933#issuecomment-1808754600

NekokeCore committed:
mediatek: add support for Ruijie RG-X60 Pro

This pull request ports Ruijie RG-X60 Pro router support to the main branch.

Parameters:
- SoC       : MediaTek MT7986A Quad-core ARM Cortex-A53 2.0GHz
- RAM       : DDR3 512MiB (W634GU6QB)
- Flash     : SPI-NAND 128 MiB (W25N01GVZEIG)
- WLAN      : MediaTek MT7986A integration dual-band WiFi 6
  - 2.4 GHz : b/g/n/ax, MIMO 4x4
  - 5 GHz   : a/n/ac/ax, MIMO 4x4
- Ethernet  : 10/100/1000 Mbps x4 (MediaTek MT7531AE)
                    2500Mbps x 1 (Realtek RTL8221B-VB-CG)
- UART      : through-hole on PCB
  - [J500] GND, TX, RX, 3.3V (115200n1)
- Buttons   : Mesh, Reset
- LEDs      : 1x Power (Blue)
                  1x Turbo (Purple)
- Power     : 12 VDC, 3 A

How to Installation:
1. Remove the case and connect the TTL cable to the corresponding position.
2. Power on the device and quickly press "down" on the keyboard, then
   U-Boot will stay in the menu.
3. Select "1. Upgrade Firmware", select "0. TFTP Client(Default)".
4. Input the IP address, input the Openwrt image file name to be
   flashed, start the TFTP server, and press "Enter".
5. Wait for the flashing to complete.

How return to stock:

1. Remove the case and connect the TTL cable to the corresponding
   position.
2. Power on the device and quickly press "down" on the keyboard, then
   U-Boot will stay in the menu.
3. Select "1. Upgrade Firmware", select "0. TFTP Client(Default)".
4. Input the IP address, input the Stock “E-WEBOS” image file name to
   be flashed, start the TFTP server, and press "Enter".
5. Wait for the flashing to complete.

About recovery:
Connect uart, use u-boot menu to flash stock firmware image or boot
OpenWrt initramfs image.

About MAC Address:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 10:82:3D:XX:XX:9E | label     |
| LAN     | 10:82:3D:XX:XX:9F | label+1   |
| WLAN 2g | 10:82:3D:XX:XX:A0 | label+2   |
| WLAN 5g | 10:82:3D:XX:XX:A1 | label+3   |
+---------+-------------------+-----------+

Signed-off-by: Ashley Lee <code@emtips.net>

NekokeCore commented:
Hi, I have corrected the PR content and commit, any other suggestions? @csharper2005

csharper2005 commented:
> Hi, I have corrected the PR content and commit

You added info to the 1st message of the PR. Commit message is still empty:
```
mediatek: add support for ruijie rg-x60-pro

Signed-off-by: Ashley Lee <code@emtips.net>

```
You can run `git commit --amend` and edit commit message.

> any other suggestions? @csharper2005

1. I don't understand how do you set correct MAC addresses (lan, wan, wlan) in OpenWrt. Please add MAC addresses table to the commit message.  An example:
```
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 24:0f:5e:xx:xx:b4 | label     |
| LAN     | 24:0f:5e:xx:xx:b5 | label+1   |
| WLAN 2g | 24:0f:5e:xx:xx:b6 | label+2   |
| WLAN 5g | 24:0f:5e:xx:xx:b7 | label+3   |
+---------+-------------------+-----------+
The WLAN 2g MAC was found in 'Factory', 0x4
```

2. Device name. RG-X60Pro, X60 Pro, RG-X60-Pro. What's exactly written on the device case label?

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -942,3 +942,13 @@ define Device/zyxel_nwa50ax-pro
+
+define Device/ruijie_rg-x60-pro
+  DEVICE_VENDOR := Ruijie
+  DEVICE_MODEL := Ruijie RG X60 Pro
```
Model usually doesn't contain vendor name. Please check the label on the device case.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=8 \
```
The max log level is `7`. Do we really need to define loglevel?
https://www.kernel.org/doc/html/latest/core-api/printk-basics.html

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+				 tearlycon=uart8250,mmio32,0x11002000";
+	};
+
+	memory {
```
```suggestion
	memory@40000000 {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+	leds-gpio {
+		compatible = "gpio-leds";
+
+		system: led-0 {
```
LED node label names usually start with `led_`.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+		compatible = "mediatek,eth-mac";
+		reg = <0>;
+		phy-mode = "2500base-x";
+		mac-address-increment = <(-1)>;
```
Remove useless `mac-address-increment = <(-1)>;`.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+		reg = <1>;
+		phy-handle = <&phy7>;
+		phy-mode = "2500base-x";
+		mac-address-increment = <1>;
```
Remove useless `mac-address-increment = <1>;`.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+	pwm0_pin: pwm0-pin-21 {
+		mux {
+			function = "pwm";
+			groups = "pwm0";
```
If you don't use these pinctrls, you don't need to define them. pwm, pcm...

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+&spi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_flash_pins>;
+	cs-gpios = <0>, <0>;
```
`cs-gpios` seems useless. Please have a check.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+				label = "Factory";
+				reg = <0x180000 0x200000>;
+				read-only;
+				nvmem-layout {
```
New line before new node.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,440 @@
+						/* 0x0000 - 0x1000 EEPROM */
+						reg = <0x0 0x1000>;
+					};
+
```
Remove unnecessary new line.

DragonBluep commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -156,6 +157,7 @@ mediatek_setup_macs()
 		wan_mac=$label_mac
 		lan_mac=$(macaddr_add "$label_mac" 1)
 		;;
+	ruijie,ruijie-rg-x60-pro|\
```
Sort it alphabetically.

DragonBluep reviewed (commented):
Please add commit message. A good example: https://github.com/openwrt/openwrt/commit/485adc9d3c436d31f9713a7d8d84adf266754e26

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,311 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 \
+				 tearlycon=uart8250,mmio32,0x11002000";
```
Maybe you want `earlycon` here?

And it's better to shrink all bootargs into one line.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,311 @@
+			color = <LED_COLOR_ID_WHITE>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 22 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "timer";
```
If you use `led-running` trigger, `linux,default-trigger` is useless.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,311 @@
+			color = <LED_COLOR_ID_PURPLE>;
+			function = LED_FUNCTION_ALARM;
+			gpios = <&pio 11 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "none";
```
All LEDs are default off. So, I think `"none"` trigger is useless.

DragonBluep commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -952,3 +952,13 @@ define Device/zyxel_nwa50ax-pro
 endef
 TARGET_DEVICES += zyxel_nwa50ax-pro
+
+define Device/ruijie_rg-x60-pro
```
Please sort it alphabetically

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,308 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1,earlycon=uart8250,mmio32,0x11002000";
```
```suggestion
		bootargs = "console=ttyS0,115200n1 earlycon=uart8250,mmio32,0x11002000";
```

NekokeCore replied:
ah,My bad，a comma out of habit.

DragonBluep commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -951,4 +961,4 @@ define Device/zyxel_nwa50ax-pro
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
-TARGET_DEVICES += zyxel_nwa50ax-pro
+TARGET_DEVICES += zyxel_nwa50ax-pro
```
Please don't remove the new line at the end.

DragonBluep commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -663,6 +663,16 @@ define Device/qihoo_360t7
 endef
 TARGET_DEVICES += qihoo_360t7
 
+define Device/ruijie_rg-x60-pro
```
`routerich_ax3000` first then `ruijie_rg-x60-pro`.

DragonBluep reviewed (commented):
Generally, LGTM. Since we changed a lot of things, please make sure it still works properly.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,308 @@
+	};
+
+	memory@40000000 {
+		reg = <0x00 0x40000000 0x00 0x10000000>;
```
I just noticed that this device has 512 MiB RAM. So 512 * 1024 * 1024 = 0x20000000.
```suggestion
		reg = <0x00 0x40000000 0x00 0x20000000>;
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,308 @@
+	status = "okay";
+};
+
+&uart2 {
```
If this device only has one uart port. Perhaps we need to trim some useless uart definitions and their pinctrls.

NekokeCore replied:
ok, i will test them

DragonBluep reviewed (commented):
This should be my final comment about this PR. Please forgive me for not finding all issues at once.

NekokeCore commented:
> This should be my final comment about this PR. Please forgive me for not finding all issues at once.

This is my first time porting Openwrt to physical hardware, thanks for your patience, I learned a lot, please forgive me for not being professional enough.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,282 @@
+	};
+};
+
+&mdio{
```
A small nit.
```suggestion
&mdio {
```

DragonBluep commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,282 @@
+};
+
+&switch {
+	ports{
```
A small nit.
```suggestion
	ports {
```

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,282 @@
+#include "mt7986a.dtsi"
+
+/ {
+	compatible = "ruijie,ruijie-rg-x60-pro","mediatek,mt7986a";
```
Do you mean "ruijie,rg-x60-pro"?

csharper2005 commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,282 @@
+	};
+
+	memory@40000000 {
+		reg = <0x00 0x40000000 0x00 0x20000000>;
```
0x0 or just 0 would be enough.

csharper2005 reviewed (commented):
Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?

Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -150,6 +151,11 @@ mediatek_setup_macs()
 		;;
+	ruijie,ruijie-rg-x60-pro)
+		wan_mac=$(mtd_get_mac_ascii product_info ethaddr)
+		lan_mac=$(macaddr_add "$lan_mac" 1)
```
Hmm... I think something is wrong in this line.

NekokeCore commented:
> Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> 
> Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`

I think i was 

> Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> 
> Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`

It seems that the wireless mac address is not found in the Factory partition.

```BASH
$ hexdump -C Factory.bin
00000000  86 79 01 00 00 0c 43 26  60 10 00 00 00 00 00 00  |.y....C&`.......|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000050  00 00 00 00 00 00 00 00  00 00 00 00 01 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000100  00 00 00 00 00 00 08 00  00 00 00 00 00 00 00 00  |................|
00000110  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000190  24 64 93 93 00 3c 00 3e  35 10 00 00 15 03 00 00  |$d...<.>5.......|
000001a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000240  00 00 00 00 00 00 00 00  00 00 00 00 a1 a7 a2 a1  |................|
00000250  a4 91 00 91 00 91 00 91  00 91 00 00 00 00 00 00  |................|
00000260  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000330  00 00 00 00 00 00 00 00  00 00 00 00 82 a0 a2 a5  |................|
00000340  a2 91 00 91 00 91 00 91  00 91 00 00 00 00 00 00  |................|
00000350  00 00 00 00 00 00 00 00  00 00 b4 00 00 00 b4 00  |................|
00000360  00 00 87 00 00 00 00 00  00 00 a4 00 b1 00 87 00  |................|
00000370  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000003e0  00 00 00 00 00 00 00 09  00 00 00 00 00 00 00 00  |................|
000003f0  00 00 00 00 00 00 03 03  03 03 03 03 03 03 03 03  |................|
00000400  03 03 03 03 03 03 03 04  04 04 03 04 04 04 03 04  |................|
00000410  04 04 03 04 04 04 00 00  00 16 1b 1b 1b 18 18 18  |................|
00000420  18 0d 00 08 0f 0d 02 09  0e 0d 02 09 0e 0d 02 09  |................|
00000430  0e 0d 02 09 0e 0d 02 09  0e 0d 02 09 0e 0d 02 09  |................|
00000440  0e 2c 2c 2c 2c 2c 2c 2c  2c 2c 2c 2c 2c 2c 2c 2c  |.,,,,,,,,,,,,,,,|
00000450  2c 2c 2c 2c 2c 2c 2c 2c  2c 00 00 00 00 00 00 00  |,,,,,,,,,.......|
00000460  00 00 00 00 00 2a 2a 2a  2a 28 28 28 28 2a 2a 2a  |.....****((((***|
00000470  2a 28 28 28 28 2a 2a 2a  2a 28 28 28 28 2a 2a 2a  |*((((****((((***|
00000480  2a 28 28 28 28 00 00 00  00 00 00 00 00 00 00 00  |*((((...........|
00000490  00 00 00 00 00 81 81 82  82 82 83 82 82 83 c1 c1  |................|
000004a0  c1 80 80 c1 c1 c1 c1 c3  c2 c1 c2 c2 c1 c2 c3 c2  |................|
000004b0  c3 c4 c4 c4 c5 80 80 c1  c1 c2 c1 c2 c3 c3 c3 87  |................|
000004c0  87 86 86 86 86 84 84 85  85 00 00 00 00 00 00 00  |................|
000004d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000540  00 7f 7f 20 25 30 36 3c  41 01 48 47 4c 52 57 5c  |... %06<A.HGLRW\|
00000550  7f 7f 7f 7f 1b 22 2a 33  3a 40 01 40 46 4c 52 5a  |....."*3:@.@FLRZ|
00000560  5e 7f 7f 00 00 00 00 00  00 00 00 00 00 00 00 00  |^...............|
00000570  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000580  00 00 00 00 00 7f 7f 7f  80 80 c5 c5 d1 d1 dd dd  |................|
00000590  e5 e5 f3 f3 40 40 80 80  cf cf db db e3 e3 ef ef  |....@@..........|
000005a0  f7 f7 40 40 80 80 c5 c5  d1 d1 d9 d9 e5 e5 ed ed  |..@@............|
000005b0  40 40 80 80 bf bf cb cb  d3 d3 df df e7 e7 40 40  |@@............@@|
000005c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000600  c0 a0 c0 e0 fc fc af e0  00 00 00 00 00 00 00 00  |................|
00000610  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|

```

csharper2005 commented:
> > Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> > Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`
> 
> It seems that the wireless mac address is not found in the Factory partition.

What about product info partition? Have you found any MACs there?

NekokeCore commented:
> > > Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> > > Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`
> > 
> > 
> > It seems that the wireless mac address is not found in the Factory partition.
> 
> What about product info partition? Have you found any MACs there?

Yes, I found the MAC for the WAN port in the product info partition, in ASCII format

csharper2005 commented:
> > > > Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> > > > Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`
> > > 
> > > 
> > > It seems that the wireless mac address is not found in the Factory partition.
> > 
> > 
> > What about product info partition? Have you found any MACs there?
> 
> Yes, I found the MAC for the WAN port in the product info partition, in ASCII format.
> 
> So, I changed 02_network and now it's the correct mac address table（I tested the compiled firmware by flashing it into the hardware and the mac address is the same as the one on the bottom of the router）

Ok, you can set correct wireless MACs using 11_fix_wifi_mac script. 
https://github.com/darknblack/openwrt/blob/4300bc668819a77226b6e8ea079d3067f216e7bb/target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac

NekokeCore commented:
> > > > > Are wireless MACs in OpenWrt the same as wireless MACs in stock firmware?
> > > > > Make sense to update the commit title too `mediatek: add support for ruijie rg-x60-pro` - >`mediatek: add support for Ruijie RG-X60 Pro`
> > > > 
> > > > 
> > > > It seems that the wireless mac address is not found in the Factory partition.
> > > 
> > > 
> > > What about product info partition? Have you found any MACs there?
> > 
> > 
> > Yes, I found the MAC for the WAN port in the product info partition, in ASCII format.
> > So, I changed 02_network and now it's the correct mac address table（I tested the compiled firmware by flashing it into the hardware and the mac address is the same as the one on the bottom of the router）
> 
> Ok, you can set correct wireless MACs using 11_fix_wifi_mac script. https://github.com/darknblack/openwrt/blob/4300bc668819a77226b6e8ea079d3067f216e7bb/target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac

ok, done

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -114,6 +114,11 @@ case "$board" in
 		[ "$PHYNBR" = "0" ] && echo "$addr" > /sys${DEVPATH}/macaddress
 		[ "$PHYNBR" = "1" ] && macaddr_add $addr 1 > /sys${DEVPATH}/macaddress
 		;;
+	ruijie,rg-x60-pro)
+		addr=$(mtd_get_mac_ascii product_info ethaddr)
+		[ "$PHYNBR" = "0" ] && macaddr_add $addr 2 > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && macaddr_add $addr 3 > /sys${DEVPATH}/macaddress
+		;;
```
Please move this block after `qihoo,360t7)` block (sorting).

csharper2005 commented:
> 
> ok, done

@NekokeCore All MAC addresses in OpenWrt are equal to the all MAC addressed in stock now. Right?

NekokeCore commented:
> > ok, done
> 
> @NekokeCore All MAC addresses in OpenWrt are equal to the all MAC addressed in stock now. Right?

After modifying 11_fix_wifi_mac, the WAN, LAN, and WIFI MAC addresses are now the same as the stock.

csharper2005 commented:
> After modifying 11_fix_wifi_mac, the WAN, LAN, and WIFI MAC addresses are now the same as the stock.

Thank you. LGTM now.

NekokeCore commented:
> > After modifying 11_fix_wifi_mac, the WAN, LAN, and WIFI MAC addresses are now the same as the stock.
> 
> Thank you. LGTM now.

You're welcome and thank you for your patience and guidance.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,283 @@
+#include "mt7986a.dtsi"
+
+/ {
+	compatible = "ruijie,rg-x60-pro","mediatek,mt7986a";
```
Missing spaces.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,283 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=7 earlycon=uart8250,mmio32,0x11002000";
```
`stdout-path = "serial0:115200n8";` is enough.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,283 @@
+		/* HSGMII 2.5GbE SoC */
+		compatible = "mediatek,eth-mac";
+		reg = <0>;
+		phy-mode = "2500base-x";
```
Leave a blank line.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,283 @@
+	phy7: phy@7 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <7>;
+		eth = <1>;
```
Useless?

NekokeCore replied:
no, ruijie have one 2.5GbE Port

aiamadeus replied:
I mean the line "eth = xxx".
Also you need to use c45 instead of c22.
```
	phy7: ethernet-phy@7 {
		compatible = "ethernet-phy-ieee802.3-c45";
		reg = <7>;
		reset-assert-us = <100000>;
		reset-deassert-us = <100000>;
		reset-gpios = <&pio 6 GPIO_ACTIVE_LOW>;
		realtek,aldps-enable;
	};
```

aiamadeus commented on target/linux/mediatek/dts/mt7986a-ruijie-rg-x60-pro.dts:
```diff
@@ -0,0 +1,283 @@
+		reset-assert-us = <100000>;
+		reset-deassert-us = <100000>;
+		reset-gpios = <&pio 6 GPIO_ACTIVE_LOW>;
+		phy-mode = "2500base-x";
```
Not needed, please remove.

NekokeCore commented:
Is there anything to do from my side to get the PR accepted? 
Or do I just have to wait for the required approval?
